### PR TITLE
Displayed Wild Categories

### PR DESCRIPTION
**Pull request by gio256**
This is a refresh of work by @emilyriehl and others that can be found [here](https://github.com/mikeshulman/Coq-HoTT/blob/1f2bff43ccfe4a392cadfcca8104e58f30ca0d75/theories/WildCat/Displayed.v). The ultimate direction is still unclear, so any thoughts to that effect are appreciated. 

A few arbitrary conventions that are up for debate:
- Where possible, anything that exists in [WildCat/Core](https://github.com/HoTT/Coq-HoTT/blob/master/theories/WildCat/Core.v) retains the same name prefixed appropriately by d (or D). For example, `Is1DCat` and `dcat_comp`.
- A single tick indicates that an object or structure lives in a displayed category over the underlying object or structure. For example, if `a : A` then `a' : D a`, and composition in displayed categories is denoted `g' $o' f'` over `g $o f`.

### CODE REVIEW

**Comment by jdchristensen** [comment]
Thanks for this, @gio256!  It looks like we should review all of these commits at once, right?  I think it would probably be good to squash these before merge, since they contain rough work along the way, but we can worry about that after review.

Can you summarize how this is different from the file Displayed.v in the tree you linked to?

**Comment by jdchristensen** [comment]
@gio256 The title says "WIP": can you say what you think still needs to be done?

File: theories/WildCat/Displayed.v [hunk 1]
```diff
@@ -0,0 +1,367 @@
...
+Proof.
+  destruct p, p'.
+  apply did.
+Defined.
+
+(* A displayed 0-functor G over a 0-functor F acts on displayed objects and
+1-cells and satisfies no axioms. *)
+Class Is0DFunctor
+  {A : Type} `{IsGraph A} {B : Type} `{IsGraph B}
+  (DA : A -> Type) `{!IsDGraph A DA} (DB : B -> Type) `{!IsDGraph B DB}
```
Comment (jdchristensen) [inline theories/WildCat/Displayed.v:60]: Because the WildCat library uses so many typeclasses, we usually let Coq insert the requirements, e.g.
```suggestion
  {A : Type} {B : Type}
  (DA : A -> Type) `{IsDGraph A DA} (DB : B -> Type) `{IsDGraph B DB}
```
The same applies to other places in this file.

**Review by jdchristensen — COMMENTED** [review]
I only skimmed the rest for now, but didn't notice any other issues.  The proofs seem to nicely follow our style.

File: theories/WildCat/Displayed.v [hunk 2]
```diff
@@ -0,0 +1,367 @@
+Require Import Basics.Overture.
+Require Import Basics.PathGroupoids.
+Require Import Basics.Tactics.
+Require Import Types.
```
[the remark this answers is not in the document]
Comment (jdchristensen) [reply theories/WildCat/Displayed.v:4]: I meant `Types.Sigma`, and that appears to be all that you need so far.

**Comment by gio256** [comment]
> Can you summarize how this is different from the file Displayed.v in the tree you linked to?

This is basically the minimum necessary changes to get that file working again and caught up to the current state of `Core.v`.

There are still a few sections of Emily's branch that aren't included here, namely functor composition and products of displayed categories. I think completing those and then revisiting makes sense.

File: theories/WildCat/Displayed.v [hunk 3]
```diff
@@ -0,0 +1,367 @@
...
+  := fun _ g' => g' $o' f'.
+
+Class Is0DGpd (A : Type) `{Is0Gpd A} (D : A -> Type)
+  `{!IsDGraph A D, !Is01DCat A D}
+  := dgpd_rev : forall {a b : A} {f : a $== b} {a' : D a} {b' : D b},
+                dhom f a' b' -> dhom (f^$) b' a'.
+
+Notation "p ^$'" := (dgpd_rev p).
+
+Definition dgpd_hom {A : Type} {D : A -> Type} `{Is0DGpd A D} {a b : A}
```
[the remark this answers is not in the document]
Comment (gio256) [reply theories/WildCat/Displayed.v:36]: Do we want to have notations for everything that `Core.v` creates notations for? I started to do this but changed my mind when I saw the length of the below.
```
a' $-[ f ]-> b'
vs.
dhom f a' b'
```

File: theories/WildCat/Displayed.v [hunk 4]
```diff
@@ -0,0 +1,367 @@
+Require Import Basics.Overture.
+Require Import Basics.PathGroupoids.
+Require Import Basics.Tactics.
+Require Import Types.
+Require Import WildCat.Core.
+
+Class IsDGraph (A : Type) `{IsGraph A} (D : A -> Type)
+  := dhom : forall {a b}, (a $-> b) -> D a -> D b -> Type.
```
[the remark this answers is not in the document]
Comment (gio256) [reply theories/WildCat/Displayed.v:8]: Because of the notational question mentioned below, `dhom` appears a lot. I am partial to `dHom` over `DHom` to minimize visual noise a bit, but I'm happy with either convention.

File: theories/WildCat/Displayed.v [hunk 5]
```diff
@@ -0,0 +1,367 @@
...
+
+Class Is01DCat (A : Type) `{Is01Cat A} (D : A -> Type) `{!IsDGraph A D} :=
+{
+  did : forall {a : A} (a' : D a), dhom (Id a) a' a';
+  dcat_comp : forall {a b c} {g : b $-> c} {f : a $-> b} {a'} {b'} {c'},
+          dhom g b' c' -> dhom f a' b' -> dhom (g $o f) a' c';
+}.
+
+Notation "g '$o'' f" := (@dcat_comp _ _ _ _ _ _ _ _ _ _ _ _ _ _ g f).
```
[the remark this answers is not in the document]
Comment (gio256) [reply theories/WildCat/Displayed.v:17]: Is it preferable to have an `Arguments` line every time to make things explicit, or is managing the implicit arguments at the definition site just as good?

[the remark this answers is not in the document]
Comment (jdchristensen) [reply theories/WildCat/Displayed.v:17]: Oh, it's fine to manage it at the site of definition, when that works.  (With Classes and Records, sometimes you want different behaviours, so you need an Arguments line.)  In this case, I see that `dcat_comp` is fine.  I was confused because you used `@dcat_comp`.  Is there a reason to not use `(dcat_comp g f)` there?

File: theories/WildCat/Displayed.v (hunk 4 above)
[the remark this answers is not in the document]
Comment (jdchristensen) [reply theories/WildCat/Displayed.v:8]: I also like `dHom`, but then I'd be inclined to use lowercase `d` everywhere, or else it will be hard to remember.

File: theories/WildCat/Displayed.v (hunk 3 above)
[the remark this answers is not in the document]
Comment (jdchristensen) [reply theories/WildCat/Displayed.v:36]: Oh, right, I forgot about the extra argument.  So maybe notations for `dHom` and `dGpdHom` aren't worth it.

**Comment by emilyriehl** [comment]
Thanks @gio256 and @jdchristensen for your careful attention to this. As I think was discussed on the Zulip, this was done ages ago when @mikeshulman invited several novice Coq users for a week-long workshop to try to make some useful contributions to the WildCats library. But despite the best efforts of our hosts/coaches, it's likely that the code in here is inelegant and buggy so please feel free to make any improvements you see fit.

File: theories/WildCat/Displayed.v (hunk 3 above)
[the remark this answers is not in the document]
Comment (Alizter) [reply theories/WildCat/Displayed.v:36]: We have used notations like `a' $-[ f ]-> b'` before but they aren't particularly useful. I think keeping them as dHom would be better. A typical example of a displayed wild category will be the displayed groupoid category of a type which has dhoms that are DPaths. Have a look in Cubical.DPath for those.

File: theories/WildCat/Displayed.v (hunk 4 above)
[the remark this answers is not in the document]
Comment (Alizter) [reply theories/WildCat/Displayed.v:8]: We went with `Hom` and `Id` because they are somewhat standard notations in category theory, but I think I agree with Dan that keeping to the lower case convention will be more useful. On the other hand, `dHom` is a family of types and we typically denote that with upper case, so I'm not sure.

[the remark this answers is not in the document]
Comment (jdchristensen) [reply theories/WildCat/Displayed.v:8]: I like the capitalization of `Hom` and `Id`.  I'm just not sure what we should do with `d` vs `D`.

[the remark this answers is not in the document]
Comment (Alizter) [reply theories/WildCat/Displayed.v:8]: Maybe its best to be consistent with DPath then.

File: theories/WildCat/Displayed.v (hunk 5 above)
[the remark this answers is not in the document]
Comment (gio256) [reply theories/WildCat/Displayed.v:17]: You're right, that is confusing. No reason for it, I will change it.

As far as `Arguments` in general, I tried to keep everything sensible, but we might need to make some adjustments once this is in use.

File: theories/WildCat/Displayed.v [hunk 6]
```diff
@@ -0,0 +1,367 @@
+Require Import Basics.Overture.
+Require Import Basics.PathGroupoids.
+Require Import Basics.Tactics.
+Require Import Types.
+Require Import WildCat.Core.
+
+Class IsDGraph (A : Type) `{IsGraph A} (D : A -> Type)
```
[the remark this answers is not in the document]
Comment (gio256) [reply theories/WildCat/Displayed.v:7]: I think I'm in favor of this, unless there's a reason we might want to consistently make the underlying category explicit.

File: theories/WildCat/Displayed.v (hunk 1 above)
Comment (jdchristensen) [reply theories/WildCat/Displayed.v]: Are there other places where you can simplify the typeclass assumptions?  You do have to be careful to not generate redundant ones, so sometimes you have to use the `!` form.

**Comment by gio256** [comment]
I think it is at feature parity with the original branch and ready for review. The big thing still on my mind is making sure that all the typeclass constraints are as minimal as possible.

File: theories/WildCat/Displayed.v (hunk 1 above)
Comment (gio256) [reply theories/WildCat/Displayed.v]: I think there likely is. I did one round of review for extraneous constraints, but could have missed some. Is there a better way to handle this than thinking through the assumptions and `Print`ing typeclasses to look for redundancies?

**Comment by jdchristensen** [comment]
@gio256 Is it ok if I push a commit or two to your branch?

**Comment by jdchristensen** [comment]
BTW, before merge, I'd be inclined to squash the first 18 commits into 1, but leave the rest to reflect changes made during review.  @gio256, we normally don't squash, so when we make a PR, we try to have the commits reflect logical steps for review.

**Comment by mikeshulman** [comment]
Thanks for doing this!  I'm very happy to see this work rescued from oblivion.

**Comment by jdchristensen** [comment]
> I'm happy to squash (or have squashed) any combination of commits, but I'm not sure I fully understand the convention. Is there something I should do here that is preferable to just [squash merging](https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/incorporating-changes-from-a-pull-request/about-pull-request-merges#squash-and-merge-your-commits) the branch?

Our githib repo is configured to not allow squash merges, because we normally like to see the history, and squash merges also don't play well with work that builds upon them.  Instead, we try to make PRs that have commits that we would want to see in the history.  In this case, once Ali confirms this is good, I can use git to squash the first 18 commits, and then force push to here before merging.

BTW, do you have plans to make additional PRs?  Do you have a particular direction you are interested in pursuing?

File: theories/WildCat/Displayed.v [hunk 7]
```diff
@@ -0,0 +1,467 @@
...
+    is0dfunctor_postcomp : forall {a b c : A} {g : b $-> c} {a' : D a}
+                            {b' : D b} {c' : D c} (g' : DHom g b' c'),
+                            Is0DFunctor (fun f => DHom f a' b')
+                                        (fun gf => DHom gf a' c')
+                                        (cat_postcomp a g) (dcat_postcomp g');
+    is0dfunctor_precomp : forall {a b c : A} {f : a $-> b} {a' : D a}
+                          {b' : D b} {c' : D c} (f' : DHom f a' b'),
+                          Is0DFunctor (fun g => DHom g b' c')
+                                      (fun gf => DHom gf a' c')
+                                      (cat_precomp c f) (dcat_precomp f');
```
Comment (Alizter) [inline theories/WildCat/Displayed.v:90]: The indentation here is a little off compared to the other fields.

File: theories/WildCat/Displayed.v [hunk 8]
```diff
@@ -0,0 +1,467 @@
...
+  := isdgraph_dhom : forall {a b} {a'} {b'},
+                      IsDGraph (fun (f:a $-> b) => DHom f a' b').
+
+Global Existing Instance isdgraph_dhom.
+#[global] Typeclasses Transparent Is2DGraph.
+
+Class Is1DCat {A : Type} `{Is1Cat A}
+  (D : A -> Type) `{!IsDGraph D, !Is2DGraph D, !Is01DCat D} := {
+    is01dcat_dhom : forall {a b : A} {a' : D a} {b' : D b},
+                    Is01DCat (fun f => DHom f a' b');
```
Comment (Alizter) [inline theories/WildCat/Displayed.v:78]: This one needs to be indented a little more to be more uniform with the others.

File: theories/WildCat/Displayed.v [hunk 9]
```diff
@@ -0,0 +1,467 @@
...
+Global Instance is0gpd_sigma {A : Type} (D : A -> Type) `{Is0DGpd A D}
+  : Is0Gpd (sig D).
+Proof.
+  srapply Build_Is0Gpd.
+  intros [a a'] [b b'] [f f'].
+  exact (f^$; dgpd_rev f').
+Defined.
+
+Global Instance is0functor_pr1 {A : Type} (D : A -> Type) `{IsDGraph A D}
+  : Is0Functor (pr1 : (sig D) -> A).
```
Comment (Alizter) [inline theories/WildCat/Displayed.v:167]:
```suggestion
  : Is0Functor (pr1 : sig D -> A).
```

File: theories/WildCat/Displayed.v [hunk 10]
```diff
@@ -0,0 +1,467 @@
...
+  (p : a = b) {a' : D a} {b': D b} (p' : transport D p a' = b')
+  : DGpdHom (GpdHom_path p) a' b'.
+Proof.
+  destruct p, p'.
+  apply DId.
+Defined.
+
+(** A displayed 0-functor [F'] over a 0-functor [F] acts on displayed objects and 1-cells and satisfies no axioms. *)
+Class Is0DFunctor {A : Type} {B : Type}
+  (DA : A -> Type) `{IsDGraph A DA} (DB : B -> Type) `{IsDGraph B DB}
```
Comment (Alizter) [inline theories/WildCat/Displayed.v:60]: I think we should make the families implicit here as they are inferrable from F'.

Comment (Alizter) [reply theories/WildCat/Displayed.v:60]: It will also make the instance parameters easier to read later.

File: theories/WildCat/Displayed.v [hunk 11]
```diff
@@ -0,0 +1,467 @@
...
+    exact (DGpdHom_path (cat_assoc_strong f g h) (dcat_assoc_strong f' g' h')).
+  - intros a b f a' b' f'.
+    exact (DGpdHom_path (cat_idl_strong f) (dcat_idl_strong f')).
+  - intros a b f a' b' f'.
+    exact (DGpdHom_path (cat_idr_strong f) (dcat_idr_strong f')).
+Defined.
+
+Class Is1DFunctor
+  {A B : Type} (DA : A -> Type) `{Is1DCat A DA} (DB : B -> Type) `{Is1DCat B DB}
+  (F : A -> B) `{!Is0Functor F, !Is1Functor F}
```
Comment (Alizter) [inline theories/WildCat/Displayed.v:291]: Same comments about the families as for Is0DFunctor

File: theories/WildCat/Displayed.v [hunk 12]
```diff
@@ -0,0 +1,467 @@
...
+    - intros a b f g p a' b' f' g' p'.
+      apply (dfmap2 _ _ (dfmap2 f' g' p')).
+    - intros a a'.
+      apply (dfmap2 _ _ (dfmap_id a') $@' dfmap_id _).
+    - intros a b c f g a' b' c' f' g'.
+      apply (dfmap2 _ _ (dfmap_comp f' g') $@' dfmap_comp _ _).
+  Defined.
+End CompositeFunctor.
+
+Local Definition obwise_prod {A B : Type} (DA : A -> Type) (DB : B -> Type)
```
Comment (Alizter) [inline theories/WildCat/Displayed.v:406]: What does obwise mean? Would fiberwise be a better name?

**Review by Alizter — APPROVED** [review]
I've left some comments.

**Comment by Alizter** [comment]
I've left some comments but I'm happy with going forward with this.

File: theories/WildCat/Displayed.v (hunk 12 above)
Comment (jdchristensen) [reply theories/WildCat/Displayed.v:406]: `obwise` is short for object-wise.  `pointwise` would be another appropriate name.  It's only used a few times, so maybe a longer name would be helpful.

File: theories/WildCat/Displayed.v (hunk 1 above)
Comment (jdchristensen) [reply theories/WildCat/Displayed.v]: I used `About` to get a more concise view, to make sure there are no duplicate superclasses.

**Comment by jdchristensen** [comment]
@gio256 You can hit the "Resolved" button as you address things, so it's clear what remains.  I'll look at this again tomorrow.

File: theories/WildCat/Displayed.v (hunk 8 above)
Comment (jdchristensen) [reply theories/WildCat/Displayed.v:78]: We don't use tab characters or tab stops.  For Class declarations, I don't think we have strong preferences about the indentation.  I'm going to push a minor change to the indentation here, and then squash the first 18 commits and merge.

**Comment by jdchristensen** [comment]
> 1. Make lemmas named something like `isgraph_equiv_isgraph` that say that if we have `A <~> B` and `IsGraph A`, then we get `IsGraph B`.  Same for the other Wildcat typeclasses.

This is already done in WildCat/Induced.v, and just needs a map (in the other direction), not an equivalence.
